File: 2022-01-28/valid-sudoku.ts
function isValidSudoku(board: string[][]): boolean {
    const rowMap = new Map();
    const colMap = new Map();
    const squareMap = new Map();
    for(let i=0;i<9;i++) {
        for(let j=0;j<9;j++) {
            
            let row=i;
            let col=j;
            if(board[row][col] !== '.') {
                if(rowMap.has(board[row][col])) return false;
                rowMap.set(board[row][col], true);
            }
            
            row=j;
            col=i;
            if(board[row][col] !== '.') {
                if(colMap.has(board[row][col])) return false;
                colMap.set(board[row][col], true);
            }
            
            row=Math.floor(i/3)*3 + Math.floor(j/3);
            col=3*Math.floor(i%3) + j%3;
            if(board[row][col] !== '.') {
                if(squareMap.has(board[row][col])) return false;
                squareMap.set(board[row][col], true);
            }
        }
        
        rowMap.clear();
        colMap.clear();
        squareMap.clear();
    }
    return true;
};

File: 2022-02-09/valid-sudoku.kt
class Solution {
    fun isValidSudoku(board: Array<CharArray>): Boolean {
        val rowSet = mutableSetOf<Char>()
        val colSet = mutableSetOf<Char>()
        val squareSet = mutableSetOf<Char>()
        for (i in 0..8) {
            for (j in 0..8) {
                val row = board[j][i]
                if (row != '.') {
                    if (rowSet.contains(row)) {
                        return false;
                    } else {
                        rowSet.add(row)
                    }
                }

                val col = board[i][j]
                if (col != '.') {
                    if (colSet.contains(col)) {
                        return false;
                    } else {
                        colSet.add(col);
                    }
                }

                val square = board[i / 3 * 3 + j / 3][i % 3 * 3 + j % 3]
                if (square != '.') {
                    if (squareSet.contains(square)) {
                        return false;
                    } else {
                        squareSet.add(square)
                    }
                }
            }
            rowSet.clear();
            colSet.clear();
            squareSet.clear();
        }

        return true;
    }
}
